File: Python/1744_수묶기.py
import sys

if __name__ == '__main__':
	N = int(sys.stdin.readline().rstrip())
	
	pos = [int(sys.stdin.readline().rstrip()) for _ in range(N)]
	pos.sort() # 입력 받은 수 오름차순으로 정렬
	
	idx = 0 # 양수와 음수 나누기 위한 인덱스(가장 작은 양수를 가리킬 것)
	for i in range(N):
		if pos[i] > 0:
			idx = i
			break
	neg = pos[:idx] # 0을 포함한 음수 담은 배열(오름 차순 정렬)
	pos = pos[idx:] # 양수 담은 배열(오름 차순 정렬)
	
	sum = 0 # 합의 최대값
	while neg: # 음수 배열이 빌 때까지
		tmp = neg.pop(0) # 절대값이 가장 큰(실제로는 가장 작은) 수부터 pop
		if neg: # 비어 있지 않다면
			tmp *= neg.pop(0) # 하나 더 pop해서 곱해준 후
		sum += tmp # 합에 더해 주기
	
	while pos: # 양수 배열이 빌 때까지
		tmp = pos.pop() # 절대값이 가장 큰(실제로도 가장 큰) 수부터 pop
		if tmp == 1: # pop한 수가 1이라면
			sum += 1 # 묶지 않고 바로 sum에 더해주고
			continue # 다음 수 pop하러
		
		if pos: # 비어 있지 않다면
			if pos[-1] == 1: # pop할 예정인 수가 1이라면
				tmp += pos.pop() # 묶지 않고 더해주기
			else: # 아니라면
				tmp *= pos.pop() # 묶어서 곱한 후
		sum += tmp # 합에 더해주기

	print(sum) # 최종 최대 합 출력

File: C++/1744_수묶기.cpp
#include <iostream>
#include <algorithm>
using namespace std;

int main() {
	int N;
	cin >> N;
	
	int neg = 0;
	int *num = new int[N];
	for(int i = 0; i < N; i++){
		cin >> num[i];
		
		if(num[i] <= 0){
			neg++;
		} // 음수의 개수(0 포함), 양수의 개수는 N에서 음수의 개수 빼면 됨
	}
	sort(num,num+N); // 오름차순으로 정렬
	
	long long int sum = 0; // 합의 최대값 담을 변수
	if(neg % 2 != 0){ // 음수의 개수가 홀수 개라면 묶어도 1개가 남을 것
		sum += num[neg-1]; // 절대값이 가장 작은(실제로는 음수 중 가장 큰) 수는 묶이지 않음 
	}
	for(int i = 0; i < neg-1; i++){ // 절대값 큰 수부터 묶어서 곱한 후 합에 더해주기(작은 수부터)
		sum += num[i] * num[i+1];
		i++;
	}
	
	// 양수의 경우, 1에 대한 처리 필요
	while(num[neg] == 1){
		sum += 1; // 1은 곱하지 않고 모두 더해주어야 함
		neg++;
	}
	if((N-neg) % 2 != 0){ // 1을 제외한 양수들의 개수가 홀수 개라면 묶어도 1개가 남을 것
		sum += num[neg]; // 절대값이 가장 작은(실제로는 양수 중 가장 작은) 수는 묶이지 않음
	}
	for(int i = N-1; i > neg; i--){ // 절대값 큰 수부터 묶어서 곱한 후 합에 더해주기(큰 수부터)
		sum += num[i] * num[i-1];
		i--;
	}
	
	cout << sum; // 계산 결과 출력
	
	return 0;
}
